tests/Helpers/JWKFetcherTest.php

```diff
@@ -0,0 +1,110 @@
+<?php
+
+namespace Auth0\Tests\Helpers;
+
+use Auth0\SDK\Helpers\JWKFetcher;
+use josegonzalez\Dotenv\Loader;
+use GuzzleHttp\Exception\RequestException;
+
+/**
+ * Class JWKFetcherTest
+ *
+ * @package Auth0\SDK\Helpers
+ */
+class JWKFetcherTest extends \PHPUnit_Framework_TestCase
+{
+
+    /**
+     * Issuer for JWKS URL.
+     *
+     * @var string
+     */
+    public static $issuer;
+
+    /**
+     * Setup before class.
+     *
+     * @return void
+     */
+    public static function setUpBeforeClass()
+    {
+        $env_path = '.env';
+        if (file_exists($env_path)) {
+            $loader = new Loader($env_path);
+            $loader->parse()->putenv(true);
+        }
+
+        self::$issuer = 'https://'.getenv('DOMAIN').'/';
+        parent::setUpBeforeClass();
+    }
+
+    /**
+     * Test that a valid JWKS is fetched.
+     *
+     * @return void
+     */
+    public function testFetchJwks()
+    {
+        $fetcher = new JWKFetcher();
+        $keys    = $fetcher->fetchKeys( self::$issuer );
+
+        $this->assertNotEmpty( $keys );
+
+        $pem_parts = $this->getPemFromKeys( $keys );
+
+        $this->assertEquals( '-----BEGIN CERTIFICATE-----', $pem_parts[0] );
+        $this->assertEquals( '-----END CERTIFICATE-----', $pem_parts[count($pem_parts) - 2] );
+    }
+
+    /**
+     * Test that a valid JWKS is fetched with an empty path passed.
+     *
+     * @return void
+     */
+    public function testFetchJwksFullUrl()
+    {
+        $fetcher = new JWKFetcher();
+        $keys    = $fetcher->fetchKeys( self::$issuer.'.well-known/jwks.json', null );
+
+        $this->assertNotEmpty( $keys );
+
+        $pem_parts = $this->getPemFromKeys( $keys );
+
+        $this->assertEquals( '-----BEGIN CERTIFICATE-----', $pem_parts[0] );
+        $this->assertEquals( '-----END CERTIFICATE-----', $pem_parts[count($pem_parts) - 2] );
+    }
+
+    /**
+     * Test that a custom JWKS URL results in a 404.
+     *
+     * @return void
+     */
+    public function testFetchJwksCustom()
+    {
+        $fetcher = new JWKFetcher();
+
+        $caught_req_exception = false;
+        try {
+            $fetcher->fetchKeys( self::$issuer, 'invalid/jwks.json' );
```

Comment:
This doesn't make sense.. why would calling the method like this cause a 404?? It looks like a valid URL is being passed. Do unit tests not integration tests.

Reply:
> why would calling the method like this cause a 404?

Because that's an invalid URL. 

> Do unit tests not integration tests.

All of the tests in this lib are integration. This class would need a re-write to make it unit test compatible (which I would love to do but is not currently a high priority).

Reply:
> Invalid URL.
```
self::$issuer = 'https://'.getenv('DOMAIN').'/';
path = 'invalid/jwks.json'

append result = 'https://'.getenv('DOMAIN').'/invalid/jwks.json';
```

`i.e. https://josh.auth0.com/invalid/jwks.json`

That looks like a good URL. That should work. Of course integration test-ly it won't since auth0 doesn't host the file there. But a user could pass his own custom domain and the path to where they are hosting it. And it should be accepted as it is still a valid scenario.